### PR TITLE
Add `distributed_utils` to expose interface for distributed ops

### LINKED ISSUE #3: Refactor `distributed_utils` module for readability and extensibility with `torch_xla`
We hope to refactor the `distributed_utils` module into a more readable and extensible `distributed` package.

The `distributed` package uses polymorphism (the base class is `DistributedStrategy` & the concrete sub-classes are `CUDAStrategy` and `XLAStrategy`) to toggle between CUDA and PT-XLA distributed utilities during runtime. The interface for all distributed ops is exposed via the `distributed_utils` module in the `distributed` package.

For more details, see b/224290413

### PR DESCRIPTION
**Pull request by thisisalbertliang**
Resolve #3 

Related to:
- `DistributedStrategy` PR: https://github.com/thisisalbertliang/training/pull/4
- `XLAStrategy` PR: https://github.com/thisisalbertliang/training/pull/5
- `CUDAStrategy` PR: https://github.com/thisisalbertliang/training/pull/6

We hope to refactor the `distributed_utils` module into a more readable and extensible `distributed` package.

The `distributed` package uses polymorphism (the base class is `DistributedStrategy` & the concrete sub-classes are `CUDAStrategy` and `XLAStrategy`) to toggle between CUDA and PT-XLA distributed utilities during runtime. The interface for all distributed ops is exposed via the `distributed_utils` module in the `distributed` package.

For more details, see b/224290413

### CODE REVIEW

**Review by miladm — APPROVED** [review]
Thanks @thisisalbertliang!

LGTM

File: image_segmentation/pytorch/runtime/distributed/distributed_utils.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
+import random
+from typing import List, Tuple
+
+import numpy as np
+import torch
+from runtime.distributed.cuda_strategy import CUDAStrategy
+from runtime.distributed.distributed_strategy import DistributedStrategy
+from runtime.distributed.xla_strategy import XLAStrategy
+
+_STRATEGY: DistributedStrategy = None
```
Comment (will-cromar) [inline image_segmentation/pytorch/runtime/distributed/distributed_utils.py]: Should this be `Optional[DistrubutedStrategy]` since it's set to None initially?

File: image_segmentation/pytorch/runtime/distributed/distributed_utils.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    global _STRATEGY
+    if flags.torch_xla:
+        _STRATEGY = XLAStrategy()
+    else:
+        _STRATEGY = CUDAStrategy()
+    return get_world_size() > 1
+
+
+def _assert_initialized(func):
+    def wrapper(*args, **kwargs):
```
Comment (will-cromar) [inline image_segmentation/pytorch/runtime/distributed/distributed_utils.py:33]: Consider using `functools.wraps` here

Explanation of what it is and why it's useful: https://stackoverflow.com/a/309000/13708606

File: image_segmentation/pytorch/runtime/distributed/distributed_utils.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    :rtype: List[int]
+    """
+    return _STRATEGY.broadcast_seeds(seeds, device)
+
+
+def setup_seeds(
+    master_seed: int, epochs: int, device: torch.device
+) -> Tuple[List[int], List[int]]:
+    """
+    Generates seeds from one master_seed.
```
Comment (will-cromar) [inline image_segmentation/pytorch/runtime/distributed/distributed_utils.py]: ultra-nit: docstrings should be spaced like this:

```
"""Generates seeds from one master_seed.

Function returns (worker_seeds, shuffling_seeds), worker_seeds are later
...
```

Relevant style guide section: https://stackoverflow.com/a/309000/13708606

File: image_segmentation/pytorch/runtime/distributed/distributed_utils.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    Seeds are generated on worker with rank 0 and broadcasted to all other
+    workers.
+
+    :param int master_seed: master RNG seed used to initialize other generators
+    :param int epochs: number of epochs
+    :param torch.device device: backend device used for distributed broadcast
+    :return: (worker_seeds, shuffling_seeds)
+    :rtype: Tuple[List[int], List[int]]
+    """
+    if master_seed == -1:
```
Comment (will-cromar) [inline image_segmentation/pytorch/runtime/distributed/distributed_utils.py]: Is `-1` a valid seed? If `-1` is valid, the input should be `Optional[int]` and you should check for `master_seed is None` to determine if you need to generate a new seed. If it is not valid, you should explain the valid range of `master_seed` in the docstring.

Also consider have the "null" seed value as the default value for the `master_seed` argument (optional)

**Review by will-cromar — CHANGES_REQUESTED** [review]

File: image_segmentation/pytorch/runtime/distributed/distributed_utils.py (hunk 1 above)
Comment (thisisalbertliang) [reply image_segmentation/pytorch/runtime/distributed/distributed_utils.py]: Sure. I have changed it to `Optional[DistributedStrategy]`

File: image_segmentation/pytorch/runtime/distributed/distributed_utils.py (hunk 2 above)
Comment (thisisalbertliang) [reply image_segmentation/pytorch/runtime/distributed/distributed_utils.py:33]: That's a great suggestion. I didn't know about `functools.wraps`. I have added it to my decorator now.

File: image_segmentation/pytorch/runtime/distributed/distributed_utils.py (hunk 3 above)
Comment (thisisalbertliang) [reply image_segmentation/pytorch/runtime/distributed/distributed_utils.py]: Looks like you pasted the wrong link by accident. 

Anyway, I have changed all my docstrings to following this spacing convention.

File: image_segmentation/pytorch/runtime/distributed/distributed_utils.py (hunk 4 above)
Comment (thisisalbertliang) [reply image_segmentation/pytorch/runtime/distributed/distributed_utils.py]: Good idea. The code for this `setup_seeds` method actually [came from the original MLPerf codebase](https://github.com/mlcommons/training/blob/1232c4f6a0262c37da5d2a895ebcaf72046b9ea7/image_segmentation/pytorch/runtime/distributed_utils.py#L52-L90). I did not make any changes to it.

Nonetheless, I think using `-1` as a default seed is indeed a bad practice. I have refactored the code to use `None` as the default seed now (see the changes I made to [`arguments.py`](https://github.com/thisisalbertliang/training/pull/7/files#diff-154155356f167e30ebd3da7dfc7157b92c70e45c9fd79c2bcb118f474d1213b6) and [`setup_seeds`](https://github.com/thisisalbertliang/training/pull/7/files#diff-36a4ec738e6548182cae59846ad64768cf934ff52ed2da35cdbbd1a9f79f6f2d)).

**Review by will-cromar — APPROVED** [review]

File: image_segmentation/pytorch/runtime/distributed/distributed_utils.py (hunk 3 above)
Comment (will-cromar) [reply image_segmentation/pytorch/runtime/distributed/distributed_utils.py]: You're right! Here's the actual link to the style guide: https://google.github.io/styleguide/pyguide.html#383-functions-and-methods
